### ISSUE TITLE
Implement the sync posts mechanism

### ISSUE THREAD

**Issue by manojmalik20** [mattermost/mattermost-plugin-msteams#14]
The mechanism for syncing posts is not yet implemented. This is also a TODO comment in the code.

#### Link
https://github.com/mattermost/mattermost-plugin-msteams-sync/blob/b9c6ab09301813f3102893f73b06930a94603095/server/plugin.go#L157

**Comment by jespino** [comment]
Actually, that is for the sync of messages on out-of-sync situation. The sync of user is implemented, but I have to improve that using subscriptions and sync mechanisms too. So I going to rename this ticket and create another one for the users part.

**Comment by manojmalik20** [comment]
@jespino Can you explain a little bit more on this issue? How will the messages be in an out-of-sync situation and how we are planning to resolve that?

**Comment by jespino** [comment]
@manojmalik20 there are multiple ways to get out of sync, but my main concern is the idea of the plugin or the mattermost server being down for some time, and then Microsoft teams stop retrying sending the messages. That can lead to a out-of-sync situation. The idea is, on reconnect try to resync whenever is possible and makes sense.

**Comment by jespino** [comment]
@ayusht2810 yes, my expectation here is not to have the ability of modify the date, my expectation is to at least be able to deliver the messages, even out of order if is needed. What we can do is, if the date/time is too far from the current date, we can add a prefix with the date/time, or something like that.

We can do some tricks to make thinking better in mattermost side, but in MSTeams the restrictions are really high.

**Comment by manojmalik20** [comment]
@jespino What are your thoughts about maintaining the edit history in both Mattermost and Teams (for compliance purposes)?

**Comment by jespino** [comment]
What do you mean about that? Mattermost is already storing the edit history, and I guess MSTeams already stores the history internally.

**Comment by manojmalik20** [comment]
I meant for this feature. Suppose the Mattermost server goes down for some time and many messages are posted and updated on MS Teams during this period of time. After the Mattermost server goes up again, our sync posts job will run and try to post all the messages that were posted or updated on Teams. But is possible that a post is edited multiple times during this period. So, when we edit messages in Mattermost, do we edit it multiple times to maintain the edit history or edit it once with the latest message?
This scenario is also applicable in vice-versa condition. (When MS Teams goes down instead of Mattermost)

**Comment by johndavidlugtu** [comment]
+1 on @jespino 's point here. I think the idea here is to lean towards keeping things simple

**Comment by ayusht2810** [comment]
@jespino @mkdbns Proposed solution to implement the sync post mechanism. [Link](https://docs.google.com/document/d/1-KotEIUS4tpCTamamJq_eVRn6_Gh-GyMpe4afppSCwo/edit)
